src/app-layer-dnp3.c

```diff
@@ -0,0 +1,2647 @@
+            goto next;
+        }
+
+        DNP3HandleUserDataResponse(dnp3, input, frame_len);
+
+    next:
+        /* Advance the input buffer. */
+        input += frame_len;
+        input_len -= frame_len;
+        processed += frame_len;
+    }
+
+    SCReturnInt(processed);
+error:
+    /* Error out. Should only happen if the data stream no longer
+     * looks like DNP3. */
+    SCReturnInt(-1);
+}
+
+/**
+ * \brief Parse incoming data.
+ *
+ * This is the entry function for DNP3 application layer data. Its
+ * main responsibility is buffering incoming data that cannot be
+ * processed.
+ *
+ * See DNP3ParseResponsePDUs for DNP3 frame handling.
+ */
+static int DNP3ParseResponse(Flow *f, void *state, AppLayerParserState *pstate,
+    uint8_t *input, uint32_t input_len, void *local_data)
+{
+    SCEnter();
+    DNP3State *dnp3 = (DNP3State *)state;
+    DNP3Buffer *buffer = &dnp3->response_buffer;
+    int processed;
+
+    if (buffer->len) {
+        if (!DNP3BufferAdd(buffer, input, input_len)) {
+            SCLogError(SC_ERR_MEM_ALLOC, "Failed to allocate memory to buffer "
+                "DNP3 response data");
+            goto error;
+        }
+        processed = DNP3HandleResponseLinkLayer(dnp3,
+            buffer->buffer + buffer->offset,
+            buffer->len - buffer->offset);
+        if (processed < 0) {
+            goto error;
+        }
+        buffer->offset += processed;
+        DNP3BufferTrim(buffer);
+    }
+    else {
+
+        /* Check if this is a banner, ignore if it is. */
+        if (DNP3ContainsBanner(input, input_len)) {
+            goto done;
+        }
+
+        processed = DNP3HandleResponseLinkLayer(dnp3, input, input_len);
+        if (processed < 0) {
+            goto error;
+        }
+        input += processed;
+        input_len -= processed;
+
+        /* Not all data was processed, buffer it. */
+        if (input_len) {
+            if (!DNP3BufferAdd(buffer, input, input_len)) {
+                SCLogError(SC_ERR_MEM_ALLOC,
+                    "Failed to allocate memory to buffer DNP3 response data");
+                goto error;
+            }
+        }
+    }
+
+done:
+    SCReturnInt(1);
+
+error:
+    /* An error occurred while processing DNP3 frames.  Dump the
+     * buffer as we can't be assured that they are valid anymore. */
+    DNP3BufferReset(buffer);
+    SCReturnInt(-1);
+}
+
+static AppLayerDecoderEvents *DNP3GetEvents(void *state, uint64_t tx_id) {
+    DNP3State *dnp3 = state;
+    DNP3Transaction *tx;
+    uint64_t tx_num = tx_id + 1;
+
+    if (dnp3->curr && dnp3->curr->tx_num == tx_num) {
+        return dnp3->curr->decoder_events;
+    }
+
+    TAILQ_FOREACH(tx, &dnp3->tx_list, next) {
+        if (tx->tx_num == tx_num) {
+            return tx->decoder_events;
+        }
+    }
+
+    return NULL;
+}
+
+static int DNP3HasEvents(void *state) {
+    SCEnter();
+    uint16_t events = (((DNP3State *)state)->events);
+    SCReturnInt(events);
```

Comment:
This is really meant to return a bool, so (events != 0) would be a better return here I think.